### PR TITLE
feat: add google sign in

### PR DESCRIPTION
**Pull request by isavega**
# Base PR
- PR based on main

# Context
- With this PR, it is possible to log in with google

# Changelog
- Create a new view for sign in
- Create a new view for sign up
- Create a new controller for omniauth callbacks
- Create a new controller for registrations
- Create a new controller for sessions
- Create a new method for user model
- Update cors

# QA
- Check if your email is on the list of testers

### CODE REVIEW

**Review by valentinjadot — CHANGES_REQUESTED** [review]
Podrias separar el PR en 2 o 3 PRs más chicos? Hay harta logica, front, gemas, etc. 🙏

**Comment by isavega** [comment]
> Podrias separar el PR en 2 o 3 PRs más chicos? Hay harta logica, front, gemas, etc. 🙏

I think that in this case it is not worth separating the PR. Actually the only important files to check are the new controllers and the User model. Otherwise, I should open 5 PR in parallel

File: .rubocop.yml [hunk 1]
```diff
@@ -17,68 +17,68 @@ AllCops:
     - rbx
   # Include common Ruby source files.
   Include:
-    - '**/*.rb'
```
Comment (rtichauerv) [inline .rubocop.yml:20]: Please replace all double quotes for singles, this way it won't show every line as changed and will be able to check the real changes (maybe there aren't)

File: Gemfile [hunk 2]
```diff
@@ -59,6 +59,10 @@ gem 'rswag-ui'
 
 gem 'rack-cors'
 
+gem 'omniauth'
```
Comment (rtichauerv) [inline Gemfile:62]: What does this gem in particular do?

File: app/controllers/users/sessions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
```
Comment (rtichauerv) [inline app/controllers/users/sessions_controller.rb:1]: Why is this needed if there are no strings defined in this file?

File: app/models/user.rb [hunk 4]
```diff
@@ -13,13 +13,20 @@
...
+         :recoverable, :rememberable, :validatable,
+         :omniauthable, omniauth_providers: [:google_oauth2]
 
   has_many :resources
   has_many :resource_comments
   has_many :resource_evaluations
   has_many :completed_learning_units
+
+  def self.from_omniauth(auth)
+    where(provider: auth.provider, uid: auth.uid).first_or_create do |user|
```
Comment (rtichauerv) [inline app/models/user.rb:26]: Is there a reason to use first_or_create instead of find_or_create_by?

File: app/views/devise/shared/_links.html.erb [hunk 5]
```diff
@@ -18,8 +32,66 @@
...
 
-<%- if devise_mapping.omniauthable? %>
-  <%- resource_class.omniauth_providers.each do |provider| %>
-    <%= link_to "Sign in with #{OmniAuth::Utils.camelize(provider)}", omniauth_authorize_path(resource_name, provider), method: :post %><br />
-  <% end %>
-<% end %>
+
+<style>
+
+  /* Devise Button */
```
Comment (rtichauerv) [inline app/views/devise/shared/_links.html.erb:38]: I think this comment is unnecessary

File: app/views/devise/shared/_links.html.erb [hunk 6]
```diff
@@ -18,8 +32,66 @@
...
+    font-family: Montserrat,sans-serif;
+    font-size: .9em;
+    padding: 15px 15px;
+    text-align: center;
+    user-select: none;
+    -webkit-user-select: none;
+    touch-action: manipulation;
+  }
+
+  /* Google Button */
```
Comment (rtichauerv) [inline app/views/devise/shared/_links.html.erb:61]: This as well

**Review by rtichauerv — CHANGES_REQUESTED** [review]

File: db/migrate/20220829133607_add_devise_to_users.rb [hunk 7]
```diff
@@ -4,7 +4,12 @@ class AddDeviseToUsers < ActiveRecord::Migration[7.0]
   def self.up
     # rubocop:disable Rails/BulkChangeTable
     change_table :users do |t|
+      t.string :name
+      t.string :uid
+      t.string :avatar_url
+      t.string :provider
       ## Database authenticatable
+
```
[the remark this answers is not in the document]
Comment (isavega) [reply db/migrate/20220829133607_add_devise_to_users.rb:12]: You are absolutely right

**Comment by rtichauerv** [comment]
I would also detail more the QA. For example, you need to run migrations, go to a specific route to test the login, another one to test the signup, maybe test signing up, then signing out and then sign in again, etc.

File: app/models/user.rb (hunk 4 above)
Comment (isavega) [reply app/models/user.rb:26]: Both could be useful, [here](https://stackoverflow.com/questions/42178674/first-or-create-vs-find-or-create-by) I leave a reference

File: app/views/devise/shared/_links.html.erb (hunk 5 above)
Comment (isavega) [reply app/views/devise/shared/_links.html.erb:38]: I like to separate the css but ok, I delete the comments

File: .rubocop.yml (hunk 1 above)
Comment (isavega) [reply .rubocop.yml:20]: I have not changed this type of configuration. I'm gonna check it!
